Urllib3: extend request hook with request body and headers

ItayGibel-helios opened the pull request:
# Description

Adding an option for an extended form of Urllib3 request hook with additional two arguments:
1. headers - a dictionary containing the request headers
2. body - a string containing the encoded request body 


## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# How Has This Been Tested?

Added a unit test named `test_extended_request_hook` testing the extended arguments

# Does This PR Require a Core Repo Change?
 - No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [X] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [X] Unit tests have been added
- [ ] Documentation has been updated

NathanielRN committed:
Worklow to publish a package from a CODEOWNER tag (#598)

* Worklow to publish a package from a CODEOWNER tag

* Prevent main script from release >=1.0 packages

* Fixes to build package workflow and script

* Better naming and remove redundant ls flag

* Add err msg for missing setup.py ls match

* Use lowercase for all non env variable vars

owais committed:
Fix tox lint command (#623)

* Fix tox lint command

- Move apt-get install command to github actions to run on an ubuntu
  system only. This keeps tox.ini cross-platform and enabled to run on
  non-ubuntu systems locally.
- Add `-y` flag to apt-get install command so it does not wait for
  configuration on CI.

* Downgrade pylint to 2.9.x

* Add pip cache to github actions cache

adamantike committed:
Allow instrumentation of newer FastAPI versions (#602)

* Allow instrumentation of newer FastAPI versions

Since the introduction of the `_instruments` runtime checks in #475, the
FastAPI instrumentation has stopped working for versions >= `0.59.0`.
However the current test suite passes even for the latest released
version at the moment (`0.67.0`).

It seems this isn't related to a limitation in the instrumentation code,
but actually because of it being created when `0.58` was the latest version:
https://github.com/open-telemetry/opentelemetry-python/commit/7bec76a22032e498911d06b9e3dab4e33528d995.

* Add changelog entry

Co-authored-by: Owais Lone <owais@users.noreply.github.com>

adriangb committed:
Make Falcon  instrumentation compatible with Falcon >=3 (#607)

* Make Falcon  instrumentation compatible with Falcon >=3

Replace falcon.api.Request -> falcon.Request which exists on both Falcon 2 & 3

* Update package.py

* Update CHANGELOG.md

* update tox.ini

* revert changes to tox.ini to test failures

* Update package.py

Co-authored-by: Owais Lone <owais@users.noreply.github.com>

owais committed:
Potential fix for flaky celery test (#629)

It seems the same in memory span exporter was being reused by multiple
tests. This change _should_ create a new instance of memeory exporter
per test/function.

Fixes #2067

owais committed:
Release/1.5.0 0.24b0 (#628)

* updating changelogs and version to 1.4.0-0.24.dev0

* updating changelogs and version to 1.5.0-0.24b0

* Update CHANGELOG.md

Co-authored-by: Leighton Chen <lechen@microsoft.com>

* Update CHANGELOG.md

Co-authored-by: Leighton Chen <lechen@microsoft.com>

Co-authored-by: Leighton Chen <lechen@microsoft.com>

owais committed:
More consistent assertion methods (#641)

* More consistent assertion methods

* Updated core SHA

ocelotl committed:
Rerun a flaky test (#639)

* Rerun a flaky test

ItayGibel-helios committed:
Urllib3: extend request hook with request body and headers

nozik commented on instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py:
```diff
@@ -92,6 +92,18 @@ def response_hook(span, request, response):
 _RequestHookT = typing.Optional[
     typing.Callable[[Span, urllib3.connectionpool.HTTPConnectionPool], None]
 ]
+_ExtendedRequestHookT = typing.Optional[
```
Why not add it to the original hook and make the new parameters optional so that calling the span with the original arguments won't fail (and so remain backwards compatible)?

ItayGibel-helios replied:
the problem is that the function itself is an argument.

so if I have an old code that pass a function with only two arguments as `request_hook`
and in the instrumentor I call this function with 4 arguments, it will raise TypeError exception

owais committed:
Improve reliability of tests (#643)

* Run tests on Windows in Github Actions

* core sha update

* format code

* fix ci yaml

* rebase

* lint

* Try without win+py3.6 fix

* Try without win+py3.6 fix

* Improve test reliability

Update some tests to use more deterministic methods of testing in memory
spans. This helps the core repo pass tests after adding Windows to CI
matrix.

ocelotl committed:
Remove unused imports (#646)

* Remove unused imports

Fixes #645

* Fix lint

* Import trace_tween_factory

* Fix lint

* Fix lint

* Fix pylint

* Try without using the symbol

* Fix isort

* Revert "Fix isort"

This reverts commit 4dbd91450b4ee099fd09ab1ca0fb8ae9c10bdcfa.

* Revert "Try without using the symbol"

This reverts commit d94f95564671a3155edf44bf71bf58f22ca309f0.

owais committed:
Added a script to generate readme file for root instrumentations (#647)

directory

This file lists down all instrumentations and the packages+versions they
support. This is helpful to find out at a glance what is exactly
supported.

Co-authored-by: alrex <aboten@lightstep.com>

nozik commented on instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_integration.py:
```diff
@@ -279,3 +279,28 @@ def response_hook(span, request, response):
+        body = "param1=1&param2=2"
+
+        pool = urllib3.HTTPConnectionPool("httpbin.org")
+        response = pool.request("GET", "/status/200", body=body, headers=headers)
```
Small comment - I'd change GET to POST so that it makes sense for the request to have a body

ItayGibel-helios committed:
Change GET to POST in test_extended_request_hook

owais committed:
Fix flaky celery docker test (#652)

The test was actually generating 2 spans but on very fast systems we
only saw one as the test ran faster than the message would travel
through the broker, trigger a task and generate 2nd span.

Switched from .delay() to .apply() so only one span is generated which
is enough for the test in question.

ItayGibel-helios committed:
added changelog entry
